### PR TITLE
Query: Many efficiency improvements to the runtime execution model.

### PR DESCRIPTION
**Pull request by anpete**
- Significant reduction in average compiled query code size
- Replace of QueryResultScopes with TransparentIdentifier struct
- Transparent identifiers only introduced when needed
- Exception interceptor no longer takes a Func
- Simplifications to query compiler
- No longer re-using parameter expressions. cc @maumar

### CODE REVIEW

File: test/EntityFramework.MicrosoftSqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs [hunk 1]
```diff
@@ -117,54 +104,48 @@ var customers
...
+                var3.Region = (string) object prm2.get_Item(10)
+                var3
                     }
                     , 
                     allowNullResult: False
                 )
             )
             , 
-            querySource: from Customer c in value(EntityQueryable`1[FunctionalTests.TestModels.Northwind.Customer]), 
+    entityAccessor: Unhandled expression type: Default
```
Comment (maumar) [inline test/EntityFramework.MicrosoftSqlServer.FunctionalTests/QueryLoggingSqlServerTest.cs]: that's a problem, I will fix this in some later checkin, the expression printer should be handling all the Expression nodes that appear in the query

**Comment by maumar** [comment]
we used to compensate for the ParameterExpression.VisitLambda re-use in the ExpressionPrinter, that code can be now removed

**Comment by anpete** [comment]
@maumar Yep, thought you would like the parameters :smile:

**Comment by maumar** [comment]
:shipit:

**Pull request merged**